services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py

```diff
@@ -0,0 +1,113 @@
+
+    assert app_settings.AUTOSCALING_EC2_ACCESS  # nosec
+    assert app_settings.AUTOSCALING_EC2_INSTANCES  # nosec
+    list_of_ec2_instances = utils_aws.get_ec2_instance_capabilities(
+        app_settings.AUTOSCALING_EC2_ACCESS, app_settings.AUTOSCALING_EC2_INSTANCES
+    )
+
+    for task in pending_tasks:
+        try:
+            ec2_instances_needed = [
+                utils_aws.find_best_fitting_ec2_instance(
+                    list_of_ec2_instances,
+                    utils_docker.get_resources_from_docker_task(task),
+                    score_type=utils_aws.closest_instance_policy,
+                )
+            ]
+            assert app_settings.AUTOSCALING_EC2_ACCESS  # nosec
+            assert app_settings.AUTOSCALING_NODES_MONITORING  # nosec
+
+            logger.debug("%s", f"{ec2_instances_needed[0]=}")
+            new_instance_dns_name = utils_aws.start_aws_instance(
+                app_settings.AUTOSCALING_EC2_ACCESS,
+                app_settings.AUTOSCALING_EC2_INSTANCES,
+                instance_type=parse_obj_as(
+                    InstanceTypeType, ec2_instances_needed[0].name
+                ),
+                tags={
+                    "io.osparc.autoscaling.created": f"{datetime.utcnow()}",
+                    "io.osparc.autoscaling.version": f"{VERSION}",
+                    "io.osparc.autoscaling.monitored_nodes_labels": json.dumps(
+                        app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_NODE_LABELS
+                    ),
+                    "io.osparc.autoscaling.monitored_services_labels": json.dumps(
+                        app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_LABELS
+                    ),
+                    "io.osparc.autoscaling.monitored_services_image_names": json.dumps(
+                        app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_IMAGE_NAMES
+                    ),
+                },
+                startup_script=await utils_docker.get_docker_swarm_join_script(),
+            )
+
+            new_node = await utils_docker.wait_for_node(new_instance_dns_name)
+            await utils_docker.tag_node(
+                new_node,
+                tags={"io.osparc.autoscaled_node": "true"}
+                | {
+                    tag_key: "true"
+                    for tag_key in app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_NODE_LABELS
+                }
+                | {
+                    tag_key: "true"
+                    for tag_key in app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_NEW_NODES_LABELS
+                },
+                available=True,
+            )
+            # NOTE: in this first trial we start one instance at a time
+            # In the next iteration, some tasks might already run with that instance
+            break
+        except Ec2InstanceNotFoundError:
+            logger.error(
```

Comment:
I guess in the next iteration of the autoscaling functionality, this will be an error kind that needs to be transmitted to the user via the frontend, right?

Reply:
right